seleniumJSEnv/src/main/scala/org/scalajs/jsenv/selenium/AbstractSeleniumJSRunner.scala

```diff
@@ -23,15 +24,30 @@ private[selenium] abstract class AbstractSeleniumJSRunner(
   protected def console: JSConsole = _console
   protected def driver: WebDriver with JavascriptExecutor = _driver
 
-  protected def startInternal(logger: Logger, console: JSConsole): Unit = synchronized {
+  protected def setupRun(logger: Logger, console: JSConsole): Unit = synchronized {
     require(_driver == null, "start() may only start one instance at a time.")
     require(_logger == null && _console == null)
     _logger = logger
     _console = console
     _driver = factory()
   }
 
-  def stop(): Unit = synchronized {
+  protected def endRun(): Try[Unit] = synchronized {
+    processConsoleLogs(console)
+
+    val errs = callPop("scalajsPopCapturedErrors").map(_.toString).toList
+
+    cleanupDriver()
+
+    if (errs.nonEmpty) {
+      val msg = ("Errors caught by browser:" :: errs).mkString("\n")
+      Failure(throw new Exception(msg))
+    } else {
+      Success(())
+    }
+  }
+
+  private def cleanupDriver(): Unit = {
```

Comment:
Yep. I'd even go as far and say that synchronizing in the finalizer is a bad idea. Because if, for whatever reason, there still is a lock around (no idea if this *can* happen), I'd much rather my finalizer to NPE than to block...